crypto/uri.c

```diff
@@ -0,0 +1,836 @@
+    }
+
+    if (p[off] != '?')
+        return 1;
+    p++;
+
+    while(span_pchars(p, &off) || span_chars(p, &off, 0, "/?"))
+        ;
+
+    /* Check if the loop stopped because of an invalid character */
+    if (!is_valid(p[off])) {
+        URIerr(URI_F_EXTRACT_QUERY, URI_R_INVALID_CHARACTER_IN_URI);
+        return 0;
+    }
+
+    if ((*query = OPENSSL_strndup(p, off)) == NULL) {
+        URIerr(URI_F_EXTRACT_QUERY, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+
+    *pp = p + off;
+    return 1;
+}
+
+/*
+ * fragment      = *( pchar / "/" / "?" )
+ */
+static int extract_fragment(char **fragment, const char **pp)
+{
+    const char *p = *pp;
+    size_t off = 0;
+
+    *fragment = NULL;
+
+    if (!is_valid(p[off])) {
+        URIerr(URI_F_EXTRACT_FRAGMENT, URI_R_INVALID_CHARACTER_IN_URI);
+        return 0;
+    }
+
+    if (p[off] != '#')
+        return 1;
+    p++;
+
+    while(span_pchars(p, &off) || span_chars(p, &off, 0, "/?"))
+        ;
+
+    /* Check if the loop stopped because of an invalid character */
+    if (!is_valid(p[off])) {
+        URIerr(URI_F_EXTRACT_FRAGMENT, URI_R_INVALID_CHARACTER_IN_URI);
+        return 0;
+    }
+
+    if ((*fragment = OPENSSL_strndup(p, off)) == NULL) {
+        URIerr(URI_F_EXTRACT_FRAGMENT, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+
+    *pp = p + off;
+    return 1;
+}
+
+/*
+ * URI           = scheme ":" hier-part [ "?" query ] [ "#" fragment ]
+ *
+ * URI-reference = URI / relative-ref
+ *
+ * relative-ref  = relative-part [ "?" query ] [ "#" fragment ]
+ */
+int OPENSSL_decode_uri(const char *uri, char **scheme, char **authority,
+                       char **path, char **query, char **fragment)
+{
+    const char *p = uri;
+
+    OPENSSL_init_crypto(0, NULL);
+
+    if (extract_scheme(scheme, &p)
+        && (*scheme
+            ? extract_hierpart(authority, path, &p)
+            : extract_relativepart(authority, path, &p))
+        && extract_query(query, &p)
+        && extract_fragment(fragment, &p)
+        && *p == '\0')
+        return 1;
+
+    if (*p != '\0') {
+        URIerr(URI_F_OPENSSL_DECODE_URI, URI_R_FAILED_TO_DECODE_URI);
+        ERR_add_error_data(3, "URI=\"", uri, "\"");
+    }
+
+    if (scheme != NULL) {
+        OPENSSL_free(*scheme);
+        *scheme = NULL;
+    }
+    if (authority != NULL) {
+        OPENSSL_free(*authority);
+        *authority = NULL;
+    }
+    if (path != NULL) {
+        OPENSSL_free(*path);
+        *path = NULL;
+    }
+    if (query != NULL) {
+        OPENSSL_free(*query);
+        *query = NULL;
+    }
+    if (fragment != NULL) {
+        OPENSSL_free(*fragment);
+        *fragment = NULL;
+    }
+
+    return 0;
+}
+
+/* Inline percent decoding */
+static ossl_inline int dehex(char c)
+{
+    return (c >= '0' && c <= '9'
+            ? c - '0'
+            : (c >= 'A' && c <= 'F'
+               ? c - 'A' + 10
+               : (c >= 'a' && c <= 'f'
+                  ? c - 'a' + 10
+                  : -1)));
```

Comment:
Agh!  No!  `OPENSSL_hexchar2int()` does a toebcdic conversion, which is *way wrong* in this case.